python/ray/data/_internal/logical/operators/map_operator.py

```diff
@@ -119,6 +120,26 @@ def __init__(
         )
 
 
+class Write(AbstractMap):
+    """Logical operator for write."""
+
+    def __init__(
+        self,
+        input_op: LogicalOperator,
+        datasource: Datasource,
+        ray_remote_args: Optional[Dict[str, Any]] = None,
+        **write_args,
+    ):
+        super().__init__(
+            "Write",
+            input_op,
+            fn=lambda x: x,
```

Comment:
This is kind of awkward that we cannot support write operator. Could we mark `fn` as optional in `AbstractMap`?

Reply:
Shouldn't fn be the write function for the data here?

Reply:
Like fn=datasource.write

Reply:
The `fn` is the UDF, and datasrouce.write is the transform_fn which is internal.
Maybe we should rename user-supplied `fn` to `udf` for clarity.

Reply:
I see. I think you actually want a hierarchy of AbstractMap>BatchMap and AbstractMap>Write then? Both are generating maps, just with different strategies.

Reply:
We don't need that hierarchy at the logical operator level IMO, that's a physical implementation detail that we're already capturing at the `LogicalOperator` -> `PhysicalOperator` planning stage. What would representing that hierarchy at the `LogicalOperator` give you?

`AbstractMap` being a base for `MapRows`, `MapBatches`, `FlatMap`, and `Filter` made sense since they all shared similar args, in particular the UDF `fn`, which the `Read` and `Write` don't have.

Reply:
Made a draft PR: https://github.com/ray-project/ray/pull/32538
Still WIP in fixing the fusion.

Reply:
Yeah, I can see it doesn't simplify planning much.

However, Write is logically a "apply func over all blocks" operation, which does seem like a type of Map. I'm not too sure why UDF is an argument of AbstractMap basically. If it wasn't, then the inheritance would be clean for Write.

Reply:
I think that's where we disagree. Logically, I think that `Write` is `Write`, that's describing the "what" of the operation. "apply func over all blocks" is the "how" we execute the operation, which is a physical plan/operator thing IMO.

In terms of `AbstractMap` remaining a useful abstraction: the only unifying args/concepts for `AbstractMap` that apply for `MapBatches`, `MapRows`, `FlatMap`, and `Filter` are:
1. The UDF `fn`,
2. The compute strategy `compute`,
3. The remote args `ray_remote_args`.

`fn` and `compute` don't apply for `Read` and `Write`, so we'd need to prune down the required bits of the `AbstractMap` data model to `ray_remote_args`, which makes it a pretty shallow abstraction. And the main advantage of the `AbstractMap` operation is that we can consolidate some of the planning logic, since we know that all ops that inherit from `AbstractMap` share some slice of their configuration, so we can share that slice of planning code for generating the physical `MapOperator`. Getting rid of `fn` and `compute` breaks that shared planning code. The better name for `AbstractMap` is probably `AbstractUDFMap` or something like that, since having a type capturing "map this UDF over blocks" is the core utility of this abstraction, since it makes planning logic easier to consolidate.

The utility of this shared planning code is probably more obvious in my operator fusion PR, where I refactor the map planner to better take advantage of this: https://github.com/ray-project/ray/blob/05a4d4d6db76423385f8310472ae9443df72f85d/python/ray/data/_internal/planner/transforms/transforms.py#L53-L67

Reply:
I'll leave it to you to decide, since I don't have much context on the fusion/planner code yet.

The only point I want to make is I don't think this class really needs to exist, except as an alias: `Write ~= MapBatches(batch_size=None, fn=datasource.write_fn, ray_remote_args={}))`. But this may just be a non-useful shortcut.

Reply:
I feel quite strongly that that's a decision for the planner to make, there's limited utility of making that translation/assertion at the logical level via logical operation abstraction hierarchies, and we should only create abstraction hierarchies at the logical operator level if it's useful for the optimization rules or the planner. That shortcut only becomes useful at the planning level, when we're creating the physical `MapOperator`.

Reply:
btw let's check if the write operator breaks the current randomize_blocks_order reordering.